### ISSUE TITLE
kexec_file_load support for device-tree blobs

### ISSUE THREAD

**Issue by klauskiwi** [antonblanchard/kexec-lite#15]
In Petitboot, we use the `--dtb argument` to pass a _patched_ devicetree blob that may contain modifications caused by the boot hooks (https://github.com/open-power/petitboot/tree/master/utils/hooks). Some of those modifications may contain an adjustment of what the console device ([linux,]stdout property of the chosen/ node) should be.

The problem is that the new kexec_file_load() method apparently doesn't support this, and from a bit of research that I did (on the ARM64 implementation that I believe we share some commonalities with), it's not completely _trivial_ to do so, because in a secure boot environment, giving complete control over the devicetree blob can be seen as a security opening (even if cmdline is as well and nobody cares about it so far).

Looks like ARM64 opted to allow the next kernel to inherit the devicetree blob from the current kernel (I think in ppc64 they all share the same memory region so that would work for us as well I guess) but before trying to change the entire architecture of how Petitboot implements dtb hooks, I was wanting to double-check that this is everyone's understanding of it as well, and what if any alternatives could be implemented.

tl;dr: kernel lockdown + kexec_file_load() broke Petitboot's ability to pass an alternative console to the next kernel, among other things.

Thanks,

 -Klaus

**Comment by erichte-ibm** [comment]
> it's not completely trivial to do so

> in a secure boot environment, giving complete control over the devicetree blob can be seen as a security opening

These were the two main problems we encountered when initially implementing the syscall. The first would require modifying the syscall to support supplying a device tree blob, which would already require more work. Then, we'd figure out how to either do that in a secure way, or still disable it in secure mode. We probably could filter certain blobs, or try to lock down what nodes can be changed/removed, but we would have to re-evaluate what in the device tree could be potentially problematic if changed (which I suspect may be quite a bit).

Unfortunately, we (the security team) not device tree or kexec experts, so I can't really provide an accurate estimate on how much effort this would require, so I am erring on the side of caution here.

Bottom line: We currently rely on the device tree being largely being unchanged from firmware in our threat model, at least as far as the secure boot-related entries are concerned. If this is something we definitely need, then that will require a lot of planning and work to implement this in a way we are happy with.

**Comment by klauskiwi** [comment]
I'm trying to understand what are our options here..

From my understanding, the target OS inherits the device-tree being used from the Skiroot environment.
That device-tree itself is dynamically built by Skiboot (versus built-in in the image itself as it happens on ARM for example)
So Skiroot (and Skiboot) already have full control over what the device-tree looks like. That should be fine, as they are considered trusted.

So feels like Skiroot already have full control over the device tree regardless, the thing is that we can't pass "another" device tree blob to the next kernel. Having that wouldn't change any aspect of the threat model I guess?

Seems like U-boot implemented something called "FIT" (Flattened Image Tree) that is really an extension of a device-tree, with kernel, initrd, and other device-trees embedded in sections (https://doc.coreboot.org/lib/payloads/fit.html). Maybe we could use something similar as an argument to kexec_file_load?

As for Petitboot, the functionality of passing a patched device-tree-blob to the next kernel seems like a standing restriction for now.. Even if Skiroot HAS complete control over the device tree in memory (that will be passed to the next kernel), I don't think there's a safe way to patch it there.

**Comment by erichte-ibm** [comment]
Just to add, kexec isn't restricted to be used by the skiroot kernel only -- we'd have to consider the implications of subsequent host kernel kexec's as well.
